### PR TITLE
doc: add command line tools

### PR DESCRIPTION
**Pull request by SantiagoTorres**
**Fixes issue #**:
None

**Description of the changes being introduced by the pull request**:

This PR adds a small series of documents to describe the current CLI tools. These are not meant to be exhaustive (i.e., to include every single command line argument), but rather to provide a little insight on how these tools can be used.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [N/A] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: doc/source/command-line-tools/in-toto-mock.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
+in-toto-mock
+============
+
+in-toto-mock is a variant of :doc:`in-toto-run` that lets you generate unsigned
+link metadata. This is useful if you, for example, want to generate link
+metadata, inspect it and sign if afterwards.
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-mock.rst]:
```suggestion
metadata, inspect it and sign it afterwards, e.g. using :doc:`in-toto-sign`.
```

File: doc/source/command-line-tools/in-toto-mock.rst [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+.. code-block:: sh
+
+    # generate the link
+    in-toto-mock -n touch -- touch file
+
+    # inspect the link
+    vi touch.link
+
+    # use in-toto sign to sign
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-mock.rst]:
```suggestion
    # use in-toto-sign to sign
```

File: doc/source/command-line-tools/in-toto-mock.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    in-toto-sign -k mykey -f touch.link
+    # link touch.xxxxxx.link is created hedre
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-mock.rst]:
```suggestion
    # link touch.xxxxxx.link is created here
```

File: doc/source/command-line-tools/in-toto-record.rst [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+Basic Usage
+-----------
+
+Like :doc:`in-toto-run`, you will need to pass a name and a cryptographic key
+parameter. In addition, you need to execute the `start` and `stop` commands to
+start recording and stop recording respectively.
+
+.. code-block:: sh
+
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-record.rst:18]: Should this also be a "Usage Example" section?

File: doc/source/command-line-tools/in-toto-run.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
+in-toto-run
+===========
+
+in-toto run is the main wrapper for in-toto to generate link metadata whlie
+carrying out a step. To do this, ir wraps any command passes and attempts to
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]:
```suggestion
carrying out a step. To do this, it wraps any command passed, and attempts to
```

File: doc/source/command-line-tools/in-toto-run.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
+in-toto-run
+===========
+
+in-toto run is the main wrapper for in-toto to generate link metadata whlie
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]:
```suggestion
in-toto run is the main wrapper for in-toto to generate link metadata while
```

File: doc/source/command-line-tools/in-toto-mock.rst [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+in-toto-mock is a variant of :doc:`in-toto-run` that lets you generate unsigned
+link metadata. This is useful if you, for example, want to generate link
+metadata, inspect it and sign if afterwards.
+
+
+Basic Usage
+-----------
+
+Given that no cryptographic signing is involved, in-toto-mock only requires you
+to pass a name paramter using ``--name`` or ``-n`` parameter. Other parameters
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-mock.rst]:
```suggestion
to pass a name parameter using ``--name`` or ``-n`` parameter. Other parameters
```

File: doc/source/command-line-tools/in-toto-run.rst [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  the step name. This is used to populate the "name" field in the link metadata
+  as well as picking up the name of the link metadata file (i.e.,
+  name.xxxx.link).
+
+Then, one key parameter. You can pass a key parameter with one of the following
+options:
+
+--key (or -k) <keyfile>:
+  Use a private key contained in the keyfile parameter. This key can be one of
+  rsa or ed25519 keys, created with :doc:`in-toto-keygen` or a PKCSv7/11 PEM
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]: Are you sure that this is PKCSv7/11?

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+its layout and the corresponding links). 
+
+If you are looking for a tool to verify signatures on individual
+metadata/layout files, take a look at :doc:`in-toto-sign` instead.
+
+
+Basic Usage
+-----------
+
+in-toto-verify only accepts to paramters: the location of the layout file and
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst]:
```suggestion
in-toto-verify only accepts two parameters: the location of the layout file and
```

File: doc/source/command-line-tools/in-toto-run.rst [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    # track all materials on the current directory 
+    # (recursively) and then products from the 
+    # build subdirectory
+    in-toto-run -k mykey -n myname -m . -p build 
+
+
+Examples
+--------
+
+Generating a pice of link metadata with no materials, but products (e.g., when
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]:
```suggestion
Generating a piece of link metadata with no materials, but products (e.g., when
```

File: doc/source/command-line-tools/in-toto-run.rst [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+.. code-block:: sh
+
+    # track all materials on the current directory 
+    # (recursively) and then products from the 
+    # build subdirectory
+    in-toto-run -k mykey -n myname -m . -p build 
+
+
+Examples
+--------
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]: Should the materials and products "example" be in the Examples section? Also, the other documents call the section "Usage Examples". Is it worth making this consistent?

File: doc/source/command-line-tools/in-toto-run.rst [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+If you want in-toto-run to track the materials and products as they are
+generated use the --materials and --products flags (or -m and -p).
+
+.. code-block:: sh
+
+    # track all materials on the current directory 
+    # (recursively) and then products from the 
+    # build subdirectory
+    in-toto-run -k mykey -n myname -m . -p build
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst:41]: The example should probably have a `-- <cmd>` argument.

File: doc/source/command-line-tools/in-toto-run.rst [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+--------
+
+Generating a pice of link metadata with no materials, but products (e.g., when
+generating files using no input):
+
+.. code-block:: sh
+
+    # track no materials, but I created foo.c, so 
+    # mark that as a product
+    in-toto-run -k mykey -n myname -p foo.c
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-run.rst]: The example should probably have a `-- <cmd>` argument.

File: doc/source/command-line-tools/in-toto-sign.rst [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+Usage Examples
+--------------
+
+Signing a layout:
+
+.. code-block:: sh
+
+    # sign a layout under mylayout.layout
+    in-toto-sign -k mykey -f mylayout.layout
+    # a signed root.layout is created
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-sign.rst:40]: This is not accurate. The signature is actually written to mylayout.layout

File: doc/source/command-line-tools/in-toto-keygen.rst [hunk 15]
```diff
@@ -0,0 +1,13 @@
+in-toto-keygen
+==============
+
+in-toto-keygen is a tool to generate and manage keys to use with other in-toto
+tooling.
+
+Basic usage
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-keygen.rst]: Similar section heads are title case. I suggest choosing a consistent case:
```suggestion
Basic Usage
```

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+.. code-block:: sh
+
+    # verify this layout and links using alice's public key
+    in-toto-verify -k alice.pub -l root.layout
+
+This will use alice's public key to verify the integrity of the layout, and
+then continue veriication following the policy specified in the layout itself.
+
+Link file location
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst:27]: Most of the section heads use title case. I think this should be consistent. I.e.:

```suggestion
Link File Location
```

File: doc/source/command-line-tools/index.rst [hunk 17]
```diff
@@ -0,0 +1,12 @@
+Command line tools
```
Comment (lukpueh) [inline doc/source/command-line-tools/index.rst]: See other comments about consistent section header case. I suggest.:

```suggestion
Command Line Tools
```

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+the location of the project owner's public key. The rest of the information
+(such as which links to load) is derived from the Layout.
+
+.. code-block:: sh
+
+    # verify this layout and links using alice's public key
+    in-toto-verify -k alice.pub -l root.layout
+
+This will use alice's public key to verify the integrity of the layout, and
+then continue veriication following the policy specified in the layout itself.
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst]:
```suggestion
then continue verification following the policy specified in the layout itself.
```

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+Link file location
+------------------
+
+Links need be in the current directory, and they must be named
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst]:
```suggestion
Links need to be in the current directory, and they must be named
```

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+:doc:`in-toto-run` and :doc:`in-toto-record` will generate link metadata named
+like this. If you require special handling of the in-toto link metadata files,
+please take a look at the library api to modify this behavior.
+
+
+Sublayouts
+----------
+
+If your layout includes sublayouts, in-toto will recurse into a subdirectory
+named `stepname.keyidprefix`, where all the links relevants to that sublayout
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst]:
```suggestion
named `stepname.keyidprefix`, where all the links relevant to that sublayout
```

File: doc/source/command-line-tools/in-toto-verify.rst [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+must exist. The sublayout itself will be contaiend where the link file usually
```
Comment (lukpueh) [inline doc/source/command-line-tools/in-toto-verify.rst]:
```suggestion
must exist. The sublayout itself will be contained where the link file usually
```

**Review by lukpueh — COMMENTED** [review]
Thanks a lot for taking a stab at this, @SantiagoTorres! This is really awesome. I left a couple of comments. Mostly minor typo fixes. 

Btw. did you take a look at the output of `in-toto-{run, record, verify, sign} --help`? There should be more usage examples that might be useful here.

**Comment by SantiagoTorres** [comment]
@lukpueh I think this is ready for another go. Thanks for the review!

One note is that I didn't know (or could think of) other usage examples for in-toto-record, so I wonder if we'd like to leave it  as-is and then add whenever inspiration hits us

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks for the changes, @SantiagoTorres! Are you sure you pushed everything? There are still many (11) unaddressed comments from my last review, including typo fix suggestions.

File: doc/source/command-line-tools/in-toto-sign.rst (hunk 14 above)
Comment (SantiagoTorres) [reply doc/source/command-line-tools/in-toto-sign.rst:40]: hmm, after re-reading the help I just noticed this. I should've cleaned my testing dir better. 

I'm starting to think this is not the most obvious behavior though. What do you think?

**Comment by lukpueh** [comment]
@SantiagoTorres, should we close here and merge in #341?

**Comment by SantiagoTorres** [comment]
we're opting for #341

**Pull request closed without merging**